Fix black window after exiting PiP, #4268

low-batt opened the pull request:
This commit will:
- Add a VideoPIPViewController class that extends PIPViewController
- This class will force drawing when entering and exiting PiP
- Change MainWindowController to use VideoPIPViewController for the PiP controller
- Remove the pendingRedrawsAfterEnteringPIP property from VideoView
- Remove the layout method from VideoView
- Remove code that set pendingRedrawsAfterEnteringPIP from the MainWindowController.enterPIP method

- [x] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [x] This implements/fixes issue #4268.

low-batt committed:
Fix black window after exiting PiP, #4268

This commit will:
- Add a VideoPIPViewController class that extends PIPViewController
- This class will force drawing when entering and exiting PiP
- Change MainWindowController to use VideoPIPViewController for the PiP
  controller
- Remove the pendingRedrawsAfterEnteringPIP property from VideoView
- Remove the layout method from VideoView
- Remove code that set pendingRedrawsAfterEnteringPIP from the
  MainWindowController.enterPIP method

uiryuu commented on iina/VideoPIPViewController.swift:
```diff
@@ -0,0 +1,35 @@
+  /// asynchronously.
+  override func viewDidLayout() {
+    super.viewDidLayout()
+    guard let controller = delegate as? MainWindowController else { return }
```
What about encapsulating these two lines into a private function?

uiryuu replied:
Also please bring back the checks for pause/album art detection

uiryuu reviewed (changes requested):
The solution LGTM, tho it need a lot of testing, especially on older systems.

low-batt committed:
Fix black window after exiting PiP, #4268

This commit will:
- Add a VideoPIPViewController class that extends PIPViewController
- This class will force drawing when entering and exiting PiP
- Change MainWindowController to use VideoPIPViewController for the PiP
  controller
- Remove the pendingRedrawsAfterEnteringPIP property from VideoView
- Remove the layout method from VideoView
- Remove code that set pendingRedrawsAfterEnteringPIP from the
  MainWindowController.enterPIP method

Updated to address review comments and change VideoPIPViewController,
adding a private forceDraw method that only forces drawing when
required.

low-batt commented:
I added a shared private function that only forces drawing when needed.

I tested the new commit under macOS 13.3 and 10.15.7 with a video file and a MP3 file with embedded album art.

HarukaMa commented:
https://user-images.githubusercontent.com/861659/229276145-2720f99d-0fa5-45c1-8d13-f8e59b172d54.mp4

When exiting PIP while the video is paused, there is still a small amount of delay (much better already as it was a blank window before). Also there is a slight flicker when the video is playing, but not sure if it's easy to fix.

low-batt commented:
This:
> Also there is a slight flicker when the video is playing, but not sure if it's easy to fix.

Is very concerning. Does this happen with IINA 1.2.0? Could this be related to #4285? What kind of Mac? Which macOS version?

I'm testing on two Macs. 

On my MacBookAir8,2 running macOS Catalina 10.15.7 I see the slight delay when exiting PiP. The picture is displayed in the main window and then a moment later it looks like the picture is painted again, this time showing the UI. Testing IINA 1.2.0 I see the same behavior except initially the main window is still showing the "This video is playing in picture in picture" screen that after a pause is refreshed with the still frame.

On my MacBookPro18,2 running macOS Ventura 13.3 I don't see exactly that same behavior with the fix or with IINA 1.2.0. PiP animates directly into the window. Some of the time I can see the main window turn black before it is refreshed with a still frame. Sometimes not.

On both machines if you use the `x` button to exit PiP it is easier to see the main window turning black before it is refreshed. This is seen with IINA 1.2.0 as well.

I've not seen the PiP flicker when playing so far.

svobs commented:
> Also there is a slight flicker when the video is playing, but not sure if it's easy to fix.

@HarukaMa To clarify, you mean "when exiting PIP while the video is playing", right?

@low-batt From a high-level perspective, I can see why there would be a noticeable flicker and it might be very difficult to prevent, because the `VideoView` is being moved from the "main window" to the "mini player window". In order to attach it to the new window, it has to first detach it from the old one. Not just that, but it has to open a new `NSWindow` in the process, which is a relatively heavyweight object compared to a view and has to hook into the greater OS in lots of ways. It's not hard to imagine that a number of things are happening in the background which are beyond our control and which could potentially cause hiccups.

I don't know why the decision was made to use a separate window for PiP. The main window already uses a `fullSizeContentView` style mask, so it could have just made modifications to the existing window and avoided a lot of issues. In fact, the code for "interactive mode" does a lot more work than PiP so that it can reuse the window, and it seems to work flawlessly by comparison [so well that it actually fixes things which are broken, like HDR! Allegedly, hah]. Unfortunately, like @HarukaMa noted, it's not an easy change.

HarukaMa commented:
> To clarify, you mean "when exiting PIP while the video is playing", right?

Yes, that's right. Not when it's playing and not being touched. I omitted that part as that's the same as the previous sentence, but I guess that still left some ambiguous around.

low-batt commented:
Ah. Thanks for clarifying. I was confused because I had recently read issue #4285 which does seem to be about video stuttering while playing in the PiP window.

It does seem like it might be possible to improve the current behavior, but the purpose of this PR is to make sure the current sources are working as well as 1.2.0 when entering and exiting PiP.

Do we think something needs to be addressed or are these changes approved?

uiryuu reviewed (approved):
I'm approving the changes since it works as good as 1.2.0
